1. Завантажити набiр зображень типу MNIST.
2. Виконати пiдготовку даних для навчання - за необхiдностi, залежно вiд
вхiдного набору.
3. Розбити данi на навчальну i перевiрочну множини у спiввiдношеннi 80%:
20%.
4. Побудувати i навчити базову модель з єдиним вихiдним шаром softmax.Для створення моделi реалiзувати власний клас. Ваги iнiцiалiзувати не-
великими випадковими значеннями, якi нормально розподiленi з нульо-
вим середнiм та одиничною дисперсiєю.
5. Побудувати глибокi моделi з кiлькома скритими шарами ReLU або tanh.
6. Моделi iз симетричними функцiями активацiї tanh навчити методом ґра-
дiєнтного спуску, використовуючи iнiцiалiзацiю ваг Глоро.
7. Моделi iз несиметричними функцiями активацiї ReLU навчити методом
ґрадiєнтного спуску, використовуючи iнiцiалiзацiю ваг Хе.
8. Для усiх моделей побудувати:
• графiки змiни значень функцiї втрат на тренувальнiй i перевiро-
чнiй множинах по мiрi навчання моделi, тобто залежно вiд кiлькостi епох,
• графiки змiни показника accuracy на тренувальнiй i перевiрочнiй
множинах по мiрi навчання моделi.
Для вiдслiдковування пiд час навчання значень функцiї втрат i пока-
зника accuracy можна використати функцiю tf.summary.scalar().
9. Пiдiбрати значення гiперпараметра швидкостi навчання. Для вiдображе-
ння кривих навчання та вказаних вище графiкiв використати TensorBoard.
10. Який метод iнiцiалiзацiї: традицiйний чи Глоро / Хе дає меншi значення
показника accuracy?
11. Чи має мiсце перенавчання побудованих моделей?
12. Зберегти контрольнi точки через фiксованi iнтервали пiд час навчан-
ня. В кiнцi навчання зберегти результуючi моделi. Вiдновити останню контрольну точку при запуску моделi, якщо навчання було перервано.
13. Обрати для нейронiв скритих шарiв рiзнi функцiї активацiї: LeakyReLU,
Parametric LeakyReLU, ELU. Спробувати з’ясувати чи впливає вибiр
функцiї активацiї на:
• час навчання мережi,
• якiсть роботи мережi на перевiрочнiй множинi,
• час надання прогнозу мережею.
14. Замiсть методу ґрадiєнтного спуску використати один з адаптивних ме-
тодiв: прискорений градiєнт Нестерова, Adagrad, Adadelta, Adam, ваги iнiцiалiзувати наведеними вище способами. Побудувати аналогiчнi гра-
фiки змiни значень функцiї втрат та змiни показника accuracy.
15. Чи покращилася якiсть навчання при використаннi адаптивного алго-
ритму оптимiзацiї?
16. Для однiєї з мереж додати нормалiзацiю за мiнi-батчами перед кожним
шаром.
• Перевiрити чи стала мережа менш чутливою до способу iнiцiалiзацiї
ваг.
• Чи прискорюється процес навчання мережi пiсля додавання такої
нормалiзацiї перед кожним шаром?
• Обрати бiльшi значення швидкостi навчання. Що можна сказати
про якiсть навчання?
• Зменшується чи збiльшується час надання прогнозу такою мере-
жею?
17. Додати до моделей один/ декiлька шарiв дропауту. Пiдiбрати значення
p - iмовiрностi вiдключення нейронiв.
• Вiдобразити у TensorBoard графiки, якi iлюструють оцiнки якостi
навчання мережi на навчальнiй та перевiрочнiй множинах. Зробити
висновки щодо якостi навчання мережi з i без використання дропа-
ута.
• Як треба змiнювати значення iмовiрностi p якщо мережа перена-
вчається?
18. Реалiзувати ще один спосiб регуляризацiї мереж - ранню зупинку навча-
ння мережi.
19. Обмежити ваги мережi, застосовуючи регуляризацiю за l1, l2 та max-
нормами.
20. Обрати найкращу модель за метриками якостi F1 (f1-score) та AUC.
21. Використати обрану модель для розпiзнавання одного iз зображень пе-
ревiрочної множини.
22. Спробувати створити нове зображення, подiбне до зображень у заданому
наборi даних. Розпiзнати це зображення за допомогою обраної найкра-
щої моделi.
23. Використовуючи багатошаровий персептрон, спробувати класифiкувати
набiр кольорових зображень типу CIFAR10 (згiдно з варiантом нижче).
Впевнитися, що багатошаровий персептрон не справляється з такою за-
дачею.

#Food Images (Food-101)

In [46]:
%load_ext tensorboard

In [36]:
import tensorflow as tf
import pandas as pd
import tensorflow_io as tfio
import cv2
import os
import random
import collections
from collections import defaultdict
from shutil import copy
from shutil import copytree, rmtree
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.image as img
import glob
import time
from collections import defaultdict
from shutil import copy
import os
import shutil
from collections import defaultdict
from tensorflow.keras.metrics import AUC
from tensorflow.keras.regularizers import l2
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten, Dense, LeakyReLU
from tensorflow.keras.initializers import RandomNormal, HeNormal
from tensorflow.keras.optimizers import Adam
from tensorflow_addons.metrics import F1Score
from tensorflow.keras.callbacks import TensorBoard, ModelCheckpoint, EarlyStopping
from tensorflow.keras.constraints import MaxNorm
from tensorflow.keras.layers import BatchNormalization, Dropout


Підключаємо набір даних та розбиваємо його на навчальний та перевірочний набори

In [4]:
import tensorflow as tf

train_directory = '/content/drive/MyDrive/food data/train'
test_directory = '/content/drive/MyDrive/food data/test'

# Define the image size and batch size
image_size = (224, 224)
batch_size = 32

# Load the training dataset with labels
train_dataset = tf.keras.preprocessing.image_dataset_from_directory(
    train_directory,
    image_size=image_size,
    batch_size=batch_size,
    label_mode='categorical'
)

# Load the testing dataset with labels
test_dataset = tf.keras.preprocessing.image_dataset_from_directory(
    test_directory,
    image_size=image_size,
    batch_size=batch_size,
    label_mode='categorical'
)

# Extract the class names
class_names = train_dataset.class_names

# Print the class names
print("Class Names:", class_names)


Found 2310 files belonging to 3 classes.
Found 700 files belonging to 3 classes.
Class Names: ['apple_pie', 'fish_and_chips', 'ice_cream']


Далі ми використовуємо послідовний клас моделі Sequential для створення порожньої моделі. Додаємо шари за допомогою add. Для компіляціїї використовуємо метод compile де задаємо оптимізатор Adam та функцію втрати. 
Для навчання використовуємо fit. Передаємо дані для тренувального та валідаційного наборів та об'єкти зворотніх зв'язів.

In [26]:
#прямого розповсюдження
model = Sequential()
model.add(Flatten(input_shape=(224, 224, 3)))
model.add(Dense(3, activation='softmax', kernel_initializer=RandomNormal(mean=0.0, stddev=1.0)))

# Compile the model
model.compile(optimizer=Adam(),
              loss=tf.losses.CategoricalCrossentropy(),
              metrics=[tf.metrics.Accuracy(), tf.metrics.AUC(), tf.metrics.Precision(), tf.metrics.Recall()])

# Define the callbacks
tensorboard_callback = TensorBoard(log_dir='./logs')
checkpoint_callback = ModelCheckpoint('./checkpoints/model_checkpoint.h5', save_best_only=True, save_weights_only=True)

# Train the model
history = model.fit(train_dataset,
                    validation_data=test_dataset,
                    callbacks=[tensorboard_callback, checkpoint_callback],
                    epochs=10)

Epoch 1/10
73/73 [==============================] - 558s 7s/step - loss: 15602.7061 - accuracy: 0.3874 - auc_4: 0.5406 - f1_score: 0.3890 - val_loss: 13569.9053 - val_accuracy: 0.4057 - val_auc_4: 0.5543 - val_f1_score: 0.4023
Epoch 2/10
73/73 [==============================] - 13s 164ms/step - loss: 11434.9346 - accuracy: 0.4394 - auc_4: 0.5795 - f1_score: 0.4416 - val_loss: 15404.8428 - val_accuracy: 0.3814 - val_auc_4: 0.5361 - val_f1_score: 0.3597
Epoch 3/10
73/73 [==============================] - 12s 159ms/step - loss: 10227.3594 - accuracy: 0.4571 - auc_4: 0.5926 - f1_score: 0.4590 - val_loss: 11815.3184 - val_accuracy: 0.4029 - val_auc_4: 0.5527 - val_f1_score: 0.3955
Epoch 4/10
73/73 [==============================] - 12s 142ms/step - loss: 9067.5078 - accuracy: 0.4671 - auc_4: 0.6003 - f1_score: 0.4693 - val_loss: 12681.2344 - val_accuracy: 0.3886 - val_auc_4: 0.5414 - val_f1_score: 0.3870
Epoch 5/10
73/73 [==============================] - 13s 162ms/step - loss: 8383.0762 - 

In [55]:
#%tensorboard --logdir=./logs

In [29]:
#Зі скритими шарами і relu
model = Sequential()
model.add(Flatten(input_shape=(224, 224, 3)))
model.add(Dense(114, activation='relu', kernel_initializer=HeNormal(), kernel_regularizer=l2(0.01)))
model.add(Dense(3, activation='softmax', kernel_initializer=HeNormal(), kernel_regularizer=l2(0.01), kernel_constraint=MaxNorm(3.0)))

# Compile the model
model.compile(optimizer=Adam(),
              loss='categorical_crossentropy',
              metrics=['accuracy', AUC(), F1Score(num_classes=3, average='macro')])

# Define the callbacks
tensorboard_callback = TensorBoard(log_dir='./logs')
checkpoint_callback = ModelCheckpoint('./checkpoints/model_checkpoint.h5', save_best_only=True, save_weights_only=True)
early_stopping_callback = EarlyStopping(patience=3)

# Train the model
history = model.fit(train_dataset,
                    validation_data=test_dataset,
                    callbacks=[tensorboard_callback, checkpoint_callback, early_stopping_callback],
                    epochs=10)

Epoch 1/10
73/73 [==============================] - 42s 546ms/step - loss: 5180.8130 - accuracy: 0.4190 - auc_5: 0.5644 - f1_score: 0.4207 - val_loss: 1380.2352 - val_accuracy: 0.4771 - val_auc_5: 0.6085 - val_f1_score: 0.3799
Epoch 2/10
73/73 [==============================] - 40s 537ms/step - loss: 1605.9432 - accuracy: 0.4797 - auc_5: 0.6106 - f1_score: 0.4814 - val_loss: 749.3631 - val_accuracy: 0.4914 - val_auc_5: 0.6189 - val_f1_score: 0.4980
Epoch 3/10
73/73 [==============================] - 40s 533ms/step - loss: 793.5861 - accuracy: 0.5009 - auc_5: 0.6264 - f1_score: 0.5036 - val_loss: 563.7888 - val_accuracy: 0.5186 - val_auc_5: 0.6386 - val_f1_score: 0.5100
Epoch 4/10
73/73 [==============================] - 41s 536ms/step - loss: 469.1257 - accuracy: 0.5563 - auc_5: 0.6676 - f1_score: 0.5590 - val_loss: 656.0915 - val_accuracy: 0.4186 - val_auc_5: 0.5641 - val_f1_score: 0.3780
Epoch 5/10
73/73 [==============================] - 39s 530ms/step - loss: 1155.9862 - accuracy: 

Далі будемо використовувати модель Sequential для побудова нейромережі.

*   Використовується активаційна функція LeakyReLU з параметром alpha=0.2.
*   Модель має два повнозв'язаних шари з 210 та 3 нейронами відповідно.
*  Використовуються ініціалізатор 'he_normal' для ініціалізації ваг та регуляризатор L2 з коефіцієнтом 0.01.

In [56]:
#%tensorboard --logdir=./logs_relu

In [38]:
#LeakyReLU
start_time = time.time()

model = Sequential()
model.add(Flatten(input_shape=(224, 224, 3)))
model.add(Dense(210))
model.add(LeakyReLU(alpha=0.2))
model.add(Dense(3, activation='softmax', kernel_initializer=HeNormal(), kernel_constraint=MaxNorm(3.0), kernel_regularizer=l2(0.01)))

model.compile(optimizer=Adam(),
              loss='categorical_crossentropy',
              metrics=['accuracy', AUC(), F1Score(num_classes=3, average='macro')])

tensorboard_callback = TensorBoard(log_dir='./logs_leaky')
checkpoint_callback = ModelCheckpoint('./checkpoints/model_leaky_checkpoint.h5', save_best_only=True, save_weights_only=True)
early_stopping_callback = EarlyStopping(patience=3)

history = model.fit(train_dataset,
                    validation_data=test_dataset,
                    callbacks=[tensorboard_callback, checkpoint_callback, early_stopping_callback],
                    epochs=10)

end_time = time.time()
execution_time = end_time - start_time
print("Execution Time:", execution_time)

Exception ignored in: <function _xla_gc_callback at 0x7f7520f10dc0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/jax/_src/lib/__init__.py", line 103, in _xla_gc_callback
    def _xla_gc_callback(*args):
KeyboardInterrupt: 


Epoch 1/10
73/73 [==============================] - 63s 838ms/step - loss: 10159.8770 - accuracy: 0.4463 - auc_11: 0.5849 - f1_score: 0.4487 - val_loss: 3869.7886 - val_accuracy: 0.3529 - val_auc_11: 0.5146 - val_f1_score: 0.2603
Epoch 2/10
73/73 [==============================] - 61s 814ms/step - loss: 2605.9797 - accuracy: 0.4671 - auc_11: 0.6002 - f1_score: 0.4677 - val_loss: 2141.3271 - val_accuracy: 0.4800 - val_auc_11: 0.6100 - val_f1_score: 0.3528
Epoch 3/10
73/73 [==============================] - 57s 765ms/step - loss: 2431.5591 - accuracy: 0.5048 - auc_11: 0.6287 - f1_score: 0.5075 - val_loss: 2294.1665 - val_accuracy: 0.4971 - val_auc_11: 0.6229 - val_f1_score: 0.3813
Epoch 4/10
73/73 [==============================] - 57s 773ms/step - loss: 1737.9578 - accuracy: 0.5186 - auc_11: 0.6389 - f1_score: 0.5215 - val_loss: 6482.9043 - val_accuracy: 0.3214 - val_auc_11: 0.4911 - val_f1_score: 0.2389
Epoch 5/10
73/73 [==============================] - 56s 757ms/step - loss: 2257.259


*   Використовується активаційна функція elu.
*   Модель має два повнозв'язаних шари з 118 та 3 нейронами відповідно.
*   Використовуються ініціалізатор 'he_normal' для ініціалізації ваг та регуляризатор L2 з коефіцієнтом 0.01.


In [54]:
#%tensorboard --logdir=./logs_leaky

In [32]:
#elu
start_time = time.time()

model = Sequential()
model.add(Flatten(input_shape=(224, 224, 3)))
model.add(Dense(118, activation='elu', kernel_initializer=HeNormal(), kernel_regularizer=l2(0.01)))
model.add(Dense(3, activation='softmax', kernel_initializer=HeNormal(), kernel_constraint=MaxNorm(3.0)))

model.compile(optimizer=Adam(),
              loss='categorical_crossentropy',
              metrics=['accuracy', AUC(), F1Score(num_classes=3, average='macro')])

tensorboard_callback = TensorBoard(log_dir='./logs_elu')
checkpoint_callback = ModelCheckpoint('./checkpoints/model_elu_checkpoint.h5', save_best_only=True, save_weights_only=True)
early_stopping_callback = EarlyStopping(patience=3)

history = model.fit(train_dataset,
                    validation_data=test_dataset,
                    callbacks=[tensorboard_callback, checkpoint_callback, early_stopping_callback],
                    epochs=10)

end_time = time.time()
execution_time = end_time - start_time
print("Execution Time:", execution_time)

Epoch 1/10
73/73 [==============================] - 42s 548ms/step - loss: 4034.8940 - accuracy: 0.4173 - auc_8: 0.5640 - f1_score: 0.4173 - val_loss: 2583.7893 - val_accuracy: 0.4286 - val_auc_8: 0.5714 - val_f1_score: 0.2060
Epoch 2/10
73/73 [==============================] - 41s 557ms/step - loss: 961.1415 - accuracy: 0.4723 - auc_8: 0.6048 - f1_score: 0.4735 - val_loss: 937.2454 - val_accuracy: 0.4586 - val_auc_8: 0.5939 - val_f1_score: 0.3100
Epoch 3/10
73/73 [==============================] - 48s 648ms/step - loss: 994.3906 - accuracy: 0.4498 - auc_8: 0.5874 - f1_score: 0.4506 - val_loss: 763.9196 - val_accuracy: 0.4114 - val_auc_8: 0.5592 - val_f1_score: 0.3913
Epoch 4/10
73/73 [==============================] - 42s 561ms/step - loss: 517.5018 - accuracy: 0.5476 - auc_8: 0.6619 - f1_score: 0.5494 - val_loss: 657.0535 - val_accuracy: 0.4971 - val_auc_8: 0.6236 - val_f1_score: 0.3852
Epoch 5/10
73/73 [==============================] - 42s 564ms/step - loss: 599.9960 - accuracy: 0.



*   Використовується активаційна функція relu.

*   Модель має три шари: два повнозв'язаних шари з 125 та 3 нейронами відповідно, а також BatchNormalization шар.
*   Використовуються ініціалізатор 'he_normal' для ініціалізації ваг, регуляризатор L2 з коефіцієнтом 0.01 та обмеження ядра MaxNorm з максимальним значенням 3.0.



In [53]:
#%tensorboard --logdir=./logs_elu

In [34]:
#ReLU 
start_time = time.time()

model = Sequential()
model.add(Flatten(input_shape=(224, 224, 3)))
model.add(Dense(125, kernel_initializer=HeNormal(), kernel_regularizer=l2(0.01), activation='relu'))
model.add(BatchNormalization())
model.add(Dense(3, activation='softmax', kernel_constraint=MaxNorm(3.0)))


# Compile the model
model.compile(optimizer=Adam(learning_rate=0),
              loss='categorical_crossentropy',
              metrics=['accuracy', AUC(), F1Score(num_classes=3, average='macro')])

# Define the callbacks
tensorboard_callback = TensorBoard(log_dir='./logs_batch')
checkpoint_callback = ModelCheckpoint('./checkpoints/model_batch_checkpoint.h5', save_best_only=True, save_weights_only=True)
early_stopping_callback = EarlyStopping(patience=3)

# Train the model
history = model.fit(train_dataset,
                    validation_data=test_dataset,
                    callbacks=[tensorboard_callback, checkpoint_callback, early_stopping_callback],
                    epochs=10)

end_time = time.time()
execution_time = end_time - start_time
print("Execution Time:", execution_time)

Epoch 1/10
73/73 [==============================] - 51s 666ms/step - loss: 3.9958 - accuracy: 0.3398 - auc_9: 0.5093 - f1_score: 0.3398 - val_loss: 4.4260 - val_accuracy: 0.3629 - val_auc_9: 0.5398 - val_f1_score: 0.3449
Epoch 2/10
73/73 [==============================] - 50s 650ms/step - loss: 3.9861 - accuracy: 0.3433 - auc_9: 0.5116 - f1_score: 0.3432 - val_loss: 4.2186 - val_accuracy: 0.3586 - val_auc_9: 0.5314 - val_f1_score: 0.3481
Epoch 3/10
73/73 [==============================] - 47s 636ms/step - loss: 4.0044 - accuracy: 0.3390 - auc_9: 0.5048 - f1_score: 0.3390 - val_loss: 4.1698 - val_accuracy: 0.3543 - val_auc_9: 0.5270 - val_f1_score: 0.3487
Epoch 4/10
73/73 [==============================] - 46s 612ms/step - loss: 3.9828 - accuracy: 0.3446 - auc_9: 0.5106 - f1_score: 0.3441 - val_loss: 4.1610 - val_accuracy: 0.3600 - val_auc_9: 0.5254 - val_f1_score: 0.3556
Epoch 5/10
73/73 [==============================] - 46s 609ms/step - loss: 3.9905 - accuracy: 0.3377 - auc_9: 0.5104

In [52]:
#%tensorboard --logdir=./logs_batch

In [37]:
#relu dropout
start_time = time.time()

model = Sequential()
model.add(Flatten(input_shape=(224, 224, 3)))
model.add(Dense(256, kernel_initializer=HeNormal(), kernel_regularizer=l2(0.01), activation='relu'))
model.add(BatchNormalization())
model.add(Dropout(0.2))
model.add(Dense(128, kernel_initializer=HeNormal(), kernel_regularizer=l2(0.01), activation='relu'))
model.add(BatchNormalization())
model.add(Dense(3, activation='softmax', kernel_constraint=MaxNorm(3.0)))

model.compile(optimizer=Adam(),
              loss='categorical_crossentropy',
              metrics=['accuracy', AUC(), F1Score(num_classes=3, average='macro')])

tensorboard_callback = TensorBoard(log_dir='./logs_batch_dropout')
checkpoint_callback = ModelCheckpoint('./checkpoints/model_batch_dropout_checkpoint.h5', save_best_only=True, save_weights_only=True)
early_stopping_callback = EarlyStopping(patience=3, verbose=1)

history = model.fit(train_dataset,
                    validation_data=test_dataset,
                    callbacks=[tensorboard_callback, checkpoint_callback, early_stopping_callback],
                    epochs=10)

end_time = time.time()
execution_time = end_time - start_time
print("Execution Time:", execution_time)

Epoch 1/10
73/73 [==============================] - 74s 974ms/step - loss: 12.1918 - accuracy: 0.4766 - auc_10: 0.6559 - f1_score: 0.4710 - val_loss: 9.1909 - val_accuracy: 0.3743 - val_auc_10: 0.5550 - val_f1_score: 0.2872
Epoch 2/10
73/73 [==============================] - 81s 1s/step - loss: 5.9841 - accuracy: 0.5108 - auc_10: 0.7039 - f1_score: 0.5082 - val_loss: 4.9561 - val_accuracy: 0.4014 - val_auc_10: 0.5722 - val_f1_score: 0.3519
Epoch 3/10
73/73 [==============================] - 73s 985ms/step - loss: 3.8023 - accuracy: 0.5195 - auc_10: 0.7089 - f1_score: 0.5149 - val_loss: 3.4645 - val_accuracy: 0.4257 - val_auc_10: 0.6151 - val_f1_score: 0.4206
Epoch 4/10
73/73 [==============================] - 72s 977ms/step - loss: 3.3357 - accuracy: 0.5216 - auc_10: 0.7155 - f1_score: 0.5182 - val_loss: 3.5522 - val_accuracy: 0.3343 - val_auc_10: 0.5510 - val_f1_score: 0.2688
Epoch 5/10
73/73 [==============================] - 72s 982ms/step - loss: 3.0472 - accuracy: 0.5290 - auc_10:

In [51]:
#%tensorboard --logdir=./logs_batch_dropout

In [69]:
from PIL import Image
image = Image.open('/content/drive/MyDrive/food data/fish2.jpg')
image = image.resize((224, 224))
image = np.array(image) / 255.0
image = np.expand_dims(image, axis=0)

In [67]:
model.save('model_elu_checkpoint.h5')

In [70]:
from tensorflow.keras.models import load_model

model = load_model('model_elu_checkpoint.h5')  
# Perform prediction
prediction = model.predict(image)

# Get the predicted class label
class_labels = ['apple-pie', 'ice-cream', 'fish_and_chips']
predicted_class_index = np.argmax(prediction)
predicted_class_label = class_labels[predicted_class_index]

print("Predicted category:", predicted_class_label)

1/1 [==============================] - 0s 110ms/step
Predicted category: ice-cream


Найкраще значення f1-store має модель elu, f1 на єпосі сягає 0.58.
Я спробувала модель elu на декількох завантажених мною фото, правильно віднесла в 1 з 3 випадків. Але це найкраща модель з усього що є